autoscaler/nanny/scaler.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nanny
+
+import (
+	"log"
+	"sort"
+	"strconv"
+)
+
+// Scaler determines the number of replicas to run
+type Scaler struct {
+	version   string
+	Verbose   bool
+	ladder    []int
+	scalerMap map[int]int32
+}
+
+func (s *Scaler) buildScalerMap(params *scalerParams) {
+	s.scalerMap = make(map[int]int32)
+	for k := range params.CoresToReplicasMap {
+		k64, err := strconv.ParseInt(k, 10, 32)
+		if err != nil {
+			log.Fatalf("Failed to parse configmap CoresToReplicaMap keys (must be integers as strings)")
+		}
+		v := params.CoresToReplicasMap[k]
+		s.scalerMap[int(k64)] = int32(v)
+	}
+	// construct a search ladder from the map
+	s.ladder = make([]int, 0, len(s.scalerMap))
+	for k := range s.scalerMap {
+		s.ladder = append(s.ladder, k)
+	}
+	sort.Ints(s.ladder)
+	for _, k := range s.ladder {
+		log.Printf("Cores GTE %5d => %3d replicas\n", k, s.scalerMap[k])
+	}
+}
+
+func (s *Scaler) scaleWithNodesAndCores(params *scalerParams, resourceVersion string, numCurrentNodes, schedulableCores int32) int32 {
+	if !(resourceVersion == s.version) {
+		// Only rebuild our tables if the configmap changed
+		log.Printf("Detected ConfigMap version change (old: %s new: %s) - rebuilding lookup tables\n", s.version, resourceVersion)
+		s.buildScalerMap(params)
+		s.version = resourceVersion
+	}
+	return s.scalerLookup(numCurrentNodes, schedulableCores)
+}
+
+func (s *Scaler) scalerLookup(numCurrentNodes, schedulableCores int32) int32 {
+	var neededReplicas int32 = 1
+	for _, coreCount := range s.ladder {
+		replicas := s.scalerMap[coreCount]
+		if int32(coreCount) > schedulableCores {
+			break
+		}
+		neededReplicas = replicas
+	}
+	// Minimum of two replicas if there are atleast 2 schedulable nodes
+	if numCurrentNodes > 1 && neededReplicas < 2 {
```

Comment:
This should not be hardcoded.  This is poilcy.

What if we had both a cores map and a nodes map, and you evaluated both and took the larger result?  That way I can say "1 replica per 32 cores" but also say "2 nodes == 2 relicas".  If they have 2 4-core VMs, they get 2 replicas.  If they have 2 64 core VMs, they get 4 replicas.